js/src/modules/_Menu.es6.js

```diff
@@ -0,0 +1,258 @@
+      throw new Error(`${msgPrefix} does not have element children`);
+    }
+    let elem = domNode.firstElementChild;
+    while (elem) {
+      const menuBarItem = elem.firstElementChild;
+      if (elem && menuBarItem && menuBarItem.tagName !== 'A') {
+        throw new Error(`${msgPrefix} has child elements are not A elements`);
+      }
+      elem = elem.nextElementSibling;
+    }
+    super(domNode);
+    this.isMenubar = true;
+  }
+
+  _createMenuItem(menuElement) {
+    return new MenubarItem(menuElement, this);
+  }
+
+  init() {
+    super.init();
+    this.domNode.setAttribute('role', 'menubar');
+    this.firstItem.setTabIndex(0);
+  }
+
+  // Set focus to a specific MenubarItem in the menu.
+  setFocusToItem(newItem) {
+    let openMenu = false;
+    // Close any existing menus.
+    this.menuItems.forEach(mbi => {
+      if (mbi.domNode.tabIndex === 0) {
+        openMenu = mbi.domNode.getAttribute('aria-expanded') === 'true';
+      }
+
+      mbi.domNode.tabIndex = -1;
+      if (mbi.popupMenu) {
+        mbi.popupMenu.close();
+      }
+    });
+
+    newItem.domNode.focus();
+    newItem.domNode.tabIndex = 0;
+
+    if (openMenu && newItem.popupMenu) {
+      newItem.popupMenu.open();
+    }
+    // Focus on the new menu, and open it if the previous menu was open.
+  }
+}
+
+export class PopupMenu extends Menu {
+  constructor(domNode, controllerObj) {
+    super(domNode);
+    this.controller = controllerObj;
+  }
+
+  _createMenuItem(menuElement) {
+    return new SubMenuItem(menuElement, this);
+  }
+
+  init() {
+    super.init();
+    this.domNode.setAttribute('role', 'menu');
+    // Add event handlers.
+    this.domNode.addEventListener('mouseover', this.handleMouseover.bind(this));
+    this.domNode.addEventListener('mouseout', this.handleMouseout.bind(this));
+  }
+
+  setFocusToItem(newItem) {
+    newItem.domNode.focus();
+  }
+
+  setFocusToController(commandParam) {
+    let command = commandParam;
+    if (typeof command !== 'string') {
+      command = '';
+    }
+    if (command === '') {
+      if (this.controller) {
+        this.controller.focusOnSelf();
+      }
+      return;
+    }
+    if (command === 'previous') {
+      this.controller.focusOnPreviousSibling();
+    } else if (command === 'next') {
+      this.controller.focusOnNextSibling();
+    }
+  }
+
+  open() {
+    // Get position and bounding rectangle of controller object's DOM node
+    const rect = this.controller.getBoundaries();
+    // Set CSS properties
+    if (!this.controller.getIsMenubarItem()) {
+      this.domNode.parentNode.style.position = 'relative';
+      this.domNode.style.display = 'block';
+      this.domNode.style.position = 'absolute';
+      this.domNode.style.left = `${rect.width}px`;
+      this.domNode.style.zIndex = '100';
+    } else {
+      this.domNode.style.display = 'block';
+      this.domNode.style.position = 'absolute';
+      this.domNode.style.top = `${rect.height - 1}px`;
```

Comment:
Why subtract a pixel here? Don’t we want it to line up exactly?

Reply:
Trying to avoid any gaps that might cause an inadvertent mouse out.

Reply:
Are there any cases where a pixel-rounding issue in a browser might affect this? If not, I’d go with not subtracting 1 pixel. I removed this on a current project since it looked obvious that it didn’t line up.